### ISSUE TITLE
Adding Autocomplete support

### ISSUE THREAD

**Issue by dsukhin** [capr/multigit#6]
I just started using mgit extensively for a project and have been making enhancements, tweaks, etc. that have helped me deploy a project that relies on mgit to me and my team members. So far, many things have been pretty specific to my use case (like scripts to load and setup all the repos with the right defaults, etc). However, now I am starting to look at more general enhancements.

One (mentioned in #5) is "mgit which \<file\>" which is just runs "mgit ls-tracked | grep \<file\> | awk '{print $1}" (one quirk is that files are always printed with respect to the project root, not the current dir so that makes things odd)
Others include warnings when adding a file to the wrong repo (by checking ls-double-tracked after "mgit <repo> add"), and things like that.
There are also some problems with using "--all" not in the root project dir (the root folder that contains the .mgit folder) (but I'm yet to figure out the root cause of that; I think it's the $pwd/.mgit section on line 509 of the git_cmd_one() function which seems to be assuming root dir)

One major enhancement I just started working on today is getting git autocomplete to work through the mgit wrapper. So far I have gotten autocomplete to list out the available repos and the available commands after typing "mgit \<TAB\>". That was pretty easy. The next trick is to invoke native git autocomplete (on "reponame") when I type "mgit <repo> \<TAB\>". That requires some trickery to have the native complete operation operate inside the respective git repo subshell and invoke the native git autocomplete.

Starting this issue to report progress, foster collaboration, and ask questions. The first question I have is where these enhancements are best placed. Autocomplete commands are usually set up (sourced) in .bashrc or by sourcing another file so perhaps this would require adding something to bashrc to make it work during setup. Would love to hear your thoughts on this.

Thanks for the amazing tool btw. Very helpful so far on my project!

**Comment by capr** [comment]
Hi there!  I'm glad you found mgit helpful. 

Thanks for the tip about using --all in a subdir (fixed in master). 

Not sure about doing ls-double-tracked on every `git add` because it takes a long time on larger projects, and also I don't think you can call that anyway if you're doing `git add` from a git subshell.

It would be very nice to have autocomplete.  If you figure it out please make a pull request. I think it's ok to add say, `mgit install-autocomplete` for installing autocomplete (I assume you use `complete -F` for that).

I think the best way to put these changes is in separate pull requests, one per feature.

**Comment by capr** [comment]
Btw, I don't suppose there's any way to add autocomplete to git.cmd too, is there? :)

**Comment by dsukhin** [comment]
Despite using windows, I'm still using bash for git so it's a slim chance I can also get autocomplete working in the cmd version. Come to think of it cmd (at least on windows 7) doesn't even have native support for tab completing more than one directory deep ... so the peospects don't look good lol)

**Comment by dsukhin** [comment]
Re: the ls-double-tracked, you bring up a very good point on speed. Perhaps I can use my new "mgit which" paradigm to do the check and avoid the O(n^2) dedup step. Btw do you prefer the name which or ls-which?

Making this dup check robust would also fixing the relative paths issue where files are always listed wrt the root dir. The double tracking check is also only a nice to have and something I hacked together. I personally haven't accidentally put things in the wrong place yet because I spent many days deciding on the separations etc. and they are super fresh in my mind but I'm thinking this would prevent someone else on my team from making a mistake. This might also be something I do as a pre commit check at the repo level; haven't decided yet.

**Comment by capr** [comment]
Pre-commit check sounds better, but how do you hook it so that it works on a git subshell? btw `mgit which` sounds good to me.

**Comment by dsukhin** [comment]
Lastly, yes, it uses complete -F for the install. I'll add the install steps into their own function in the mgit file and I guess the README can let people know to add mgit install-autocomplete to their bashrc so that it gets invoked at every shell run.

So far, in order to pass through git's autocomplete (without directly hooking into their completion functions, a beast of a problem on its own because of how they organized it!) I've been resorting to a subshell trick described deep in the comments here: https://stackoverflow.com/questions/24514143/bash-getting-a-commands-completion-output-programmatically-e-g-in-a-variabl#comment38050661_24514947
It makes the autocomplete lag a little more than I would like (because I am literally parsing raw bash output) so I'm open to ideas :)

**Comment by dsukhin** [comment]
I don't have it work in the subshell. I think you are right, there wouldn't be a way to do that in any elegant manner except writing an interpreter layer that catches commands we want to hook like add and doinging work outside the subshell in parallel. I haven't used the subshell much. It's a little wonky on git-bash on windows and with autocomplete working already on the repo names for me, it's almost quicker to just type "mgit repo..." for every change instead of entering and exiting the subshell.

**Comment by dsukhin** [comment]
Just did some major digging and reverse engineering and figured out how to hook the git autocomplete functions directly instead of using that nasty bash trick I referenced above. 

Working out a few quirks then I'll package it into mgit, add the requisite error checking, and submit a PR.

**Comment by dsukhin** [comment]
just added the "which" PR.
next, I was moving the working autocomplete code from my .bashrc into mgit but seems it will have to live as a separate file that needs to be sourced in bashrc (or via an mgit enable-autocomplete shortcut) otherwise it can't be properly referenced by the complete command. Let me know if you can think of a better way.

**Comment by capr** [comment]
Merged. Thanks.
